README.md
# Node-Srs

Linux: [![Build Status](https://secure.travis-ci.org/mapbox/node-srs.svg)](http://travis-ci.org/mapbox/node-srs)

Windows: [![Build status](https://ci.appveyor.com/api/projects/status/ff1n3h7h506i4vx0)](https://ci.appveyor.com/project/Mapbox/node-srs)

This module tries to detect projections, also known as "spatial reference systems". It works similiarly to [gdalsrsinfo](http://www.gdal.org/gdalsrsinfo.html).

`node-srs` supports parsing a variety of textual representations of projections, like the formats known as `OGC WKT`, `ESRI WKT`, `OGC CRS URN`, or `proj4`. It supports [shapefiles](http://en.wikipedia.org/wiki/Shapefile) and [GeoJSON](http://geojson.org/). Shapefiles optionally come with a separate .prj and inside the `.prj` the text is usually in the `ESRI WKT` format. GeoJSON optionally contains a `crs` property that declares projection as `OGC CRS URN`.

Detecting projections is important for applications like TileMill, which - through Mapnik - needs the `proj4` representation of a projection to create coordinate transformations for re-projecting vector or raster data on the fly.

`node-srs` includes a variety of hacks to determine if your projection looks like `web mercator` (epsg:3857) or `wgs84` (epsg:4326) and if so returns the canonical representations of these projections (according to @springmeyer). This ensures applications like TileMill can avoid unneeded projection. It is common for data out in the wild in web mercator projection to store slightly different projection strings based on the software that created the files. `node-srs` ensures a consistent final `proj4` representation is returned for all of the variety of representations of web mercator. This is critical because even mercator to mercator transformations are expensive if proj4 is asked to do this for large datasets.

`node-srs` does not support looking for, or detecting, projection information in formats like GeoTIFF, PostGIS, or SQLite. Rather for those formats you would need to extract the projection information yourself and then pass it to `node-srs`.

## API

### `srs.parse(string)`

Parse a string of projection specification data and return an object describing
the detected projection. If the SRS cannot be parsed, throws a `TypeError`
describing the issue.

## Example

Detect a proj4 literal string as spherical mercator:

```js
> var srs = require('srs');
> srs.parse('+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs')
{ proj4: '+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over',
  srid: 3857,
  auth: 'EPSG',
  pretty_wkt: 'PROJCS["WGS 84 / Pseudo-Mercator",\n    GEOGCS["WGS 84",\n        DATUM["WGS_1984",\n            SPHEROID["WGS 84",6378137,298.257223563,\n                AUTHORITY["EPSG","7030"]],\n            AUTHORITY["EPSG","6326"]],\n        PRIMEM["Greenwich",0,\n            AUTHORITY["EPSG","8901"]],\n        UNIT["degree",0.0174532925199433,\n            AUTHORITY["EPSG","9122"]],\n        AUTHORITY["EPSG","4326"]],\n    UNIT["metre",1,\n        AUTHORITY["EPSG","9001"]],\n    PROJECTION["Mercator_1SP"],\n    PARAMETER["central_meridian",0],\n    PARAMETER["scale_factor",1],\n    PARAMETER["false_easting",0],\n    PARAMETER["false_northing",0],\n    EXTENSION["PROJ4","+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs"],\n    AUTHORITY["EPSG","3857"],\n    AXIS["X",EAST],\n    AXIS["Y",NORTH]]',
  esri: false,
  name: 'Google Maps Global Mercator',
  valid: true,
  is_geographic: false,
  input: '+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs' }
```

Detect a WKT string as WGS84:

```js
> srs.parse('GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137,298.257223563]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]]')
{ input: 'GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137,298.257223563]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]]',
  proj4: '+proj=longlat +ellps=WGS84 +no_defs',
  srid: 4326,
  auth: 'EPSG',
  pretty_wkt: 'GEOGCS["GCS_WGS_1984",\n    DATUM["D_WGS_1984",\n        SPHEROID["WGS_1984",6378137,298.257223563]],\n    PRIMEM["Greenwich",0],\n    UNIT["Degree",0.017453292519943295],\n    AUTHORITY["EPSG","4326"]]',
  esri: false,
  name: 'GCS_WGS_1984',
  valid: true,
  is_geographic: true }
```

## Installation

    npm install

## Test

    npm test

## License

  BSD, see LICENSE.txt


CHANGELOG.md
# changelog

## 1.2.0

 - Upgraded to node-gdal@0.9.x
 - Test on OSX
 - Test with Node.js v4, v5, v6

## 1.1.0

 - Upgraded to node-gdal@0.8.0

## 1.0.1

 - Removed some unnecessary data from package

## 1.0.0

 - Re-implemented as a JS module on top of node-gdal.

## 0.4.9

 - Upgraded to gdal 2.0
 - Now building against libc++ on osx
 - Now providing binaries against Visual Studio 2015 (pass --toolset=v140) and use binaries from https://github.com/mapbox/node-cpp11


## 0.4.8

 - Upgraded to nan@1.8.4
 - Upgraded to node-pre-gyp@0.6.7

0.4.7

 - Added support (and pre-build binaries) for io.js 1.2.x and node v0.12.x
 - Upgraded to nan@1.7.0
 - Upgraded to node-pre-gyp@0.6.4

0.4.6

 - Removed getenv workaround for Visual Studio 2013

## 0.4.5

 - Upgraded bundled node-pre-gyp

## 0.4.4

 - Upgraded to GDAL 1.11.1
 - Stopped providing binaries for node v0.8.x
 - Now providing binaries against Visual Studio 2014 (pass --toolset=v140) and use binaries from https://github.com/mapbox/node-cpp11

## 0.4.3

 - No changes: just re-published to try to avoid npm shasum error upon download

## 0.4.2

 - Now supporting node v0.11.x

## 0.4.1

 - Now shipping binaries for OS X, Windows, and Ubuntu Linux

## 0.4.0

 - Now builds on FreeBSD
 - `srs.jsonCrs` added for detecting projection inside JSON object representing GeoJSON. Pass the result to `srs.parse`
 - `srs.parse` no longer supports being passed a filepath to a GeoJSON.

## 0.3.12

 - Fixed generation of valid `srs_settings.js` on windows (#39)
 - Fixed detection of some ESRI variant projections (#38)
 - Added preliminary appveyor.yml for continuous builds on windows

## 0.3.11

 - Removed binary stripping on OS X which may cause link problems on mavericks
 - Move to Mapbox organization

## 0.3.10

 - Improved support for topojson

## 0.3.9

 - `--runtime_link` option fixed to only apply when `--shared_gdal` is passed.

## 0.3.8

 - Build fixes for windows

## 0.3.7

 - Build fix for python3 (#33)

## 0.3.6

 - Build fixes to avoid possible undefined symbol errors are runtime on ubuntu linux
 - Additional travis.ci testing of 32bit builds

## 0.3.5

 - Minor test fixes to work with variable external gdal versions

## 0.3.4

 - Changed name of build option to configure against shared gdal lib. Now pass `npm install --shared_gdal` (to be consistent with older node-srs).

## 0.3.3

 - Now building against internal `osr` again by default. Pass `npm install --gdal=shared` to build against systemwide GDAL (#30)
 - Now forcing canonical wgs84/epsg:4326 represenation
 - Now translating `+init` syntax to `+proj` for known projections (epsg:4326 and epsg:3857)
 - Various fixes to detect more projections

## 0.3.2

 - Re-enabled optional linking with `gdal-config --dep-libs` by passing `npm install --runtime_link=static`

## 0.3.1

 - Removed build linking to gdal libs / gdal-config --dep-libs, now only linking to libgdal itself by default

## 0.3.0

 - Now using node-gyp for build
 - Now requiring Node >= 0.6.13 (for node-gyp support)
 - Node v0.10.x support
 - Now requiring external libgdal
 - Better detection of more spherical mercator variants
 - Better detection of +init=epsg based mercator srs and auto-transformation to +proj syntax


lib/constants.js
/*
 This is a custom srs definition for
 spherical mercator that we make
 canonical because no single variant
 of the projection stored by various
 applications is optimal.

 This is basically a modification of gdal's
 understanding of epsg 3857

 Changes:

 * add +over and +wktext to proj4 string
 * remove any double spaces in proj4 string
 * epsg is 3857, the new official epsg for 900913
 * name is pulled from gdal/data/cubwerx_extra.wkt for 900913
   instead of using 'WGS 84 / Pseudo-Mercator' or 'Popular...'

 TODO:
 * +x_0=0.0 vs +x_0=0 (same with +y_0)
 * handle two scaling aliases +k and +k_0
 * meaning of +k=1. vs +k=1.0 ?

*/
var merc_pretty_wkt = 'PROJCS["WGS 84 / Pseudo-Mercator",\n    GEOGCS["WGS 84",\n        DATUM["WGS_1984",\n            SPHEROID["WGS 84",6378137,298.257223563,\n                AUTHORITY["EPSG","7030"]],\n            AUTHORITY["EPSG","6326"]],\n        PRIMEM["Greenwich",0,\n            AUTHORITY["EPSG","8901"]],\n        UNIT["degree",0.0174532925199433,\n            AUTHORITY["EPSG","9122"]],\n        AUTHORITY["EPSG","4326"]],\n    UNIT["metre",1,\n        AUTHORITY["EPSG","9001"]],\n    PROJECTION["Mercator_1SP"],\n    PARAMETER["central_meridian",0],\n    PARAMETER["scale_factor",1],\n    PARAMETER["false_easting",0],\n    PARAMETER["false_northing",0],\n    EXTENSION["PROJ4","+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over"],\n    AUTHORITY["EPSG","3857"],\n    AXIS["X",EAST],\n    AXIS["Y",NORTH]]';

var lon_lat_pretty_wkt = 'GEOGCS["WGS 84",\n    DATUM["WGS_1984",\n        SPHEROID["WGS 84",6378137,298.257223563,\n            AUTHORITY["EPSG","7030"]],\n        TOWGS84[0,0,0,0,0,0,0],\n        AUTHORITY["EPSG","6326"]],\n    PRIMEM["Greenwich",0,\n        AUTHORITY["EPSG","8901"]],\n    UNIT["degree",0.0174532925199433,\n        AUTHORITY["EPSG","9108"]],\n    AUTHORITY["EPSG","4326"]]';

module.exports.merc_names = [
    'Google_Maps_Global_Mercator',
    'Google Maps Global Mercator',
    'WGS 84 / Pseudo-Mercator', // 3857
    'WGS_84_Pseudo_Mercator', // 3857 with new gdal/ogr name as per https://github.com/mapbox/node-srs/issues/27
    'Popular Visualisation CRS / Mercator', // 3785
    'WGS_1984_Web_Mercator', // ESRI
    'WGS_1984_Web_Mercator_Auxiliary_Sphere', // ESRI
    'Popular_Visualisation_CRS_Mercator_deprecated', // wtf
    'Mercator_1SP',
    'Mercator_2SP'
];

module.exports.merc_srids = [
    900913, // http://crschmidt.net/blog/archives/243/google-projection-900913/
    3857, // epsg official, second try
    3785, // epsg deprecated, failed first try
    102100, // esri official, second try
    102113, // esri deprecated, failed first try
    41001 //osgeo
];

// commonly confused as spherical mercator, but not, right?
// here only for reference currently, nothing done programmatically
module.exports.other_merc_srids = [
    3395, // "WGS 84 / World Mercator" or normal, no googlized mercator: lacked +a=6378137 +b=6378137
    3587, // common typo but actually NAD83(NSRS2007) / Michigan Central
    // for these see also notes in shapefile.3857.test.js
    9804, // older, esri style-named traditional merc: http://www.remotesensing.org/geotiff/proj_list/mercator_1sp.html
    9805, // older, esri style-named traditional merc: http://www.remotesensing.org/geotiff/proj_list/mercator_2sp.html
];

module.exports.known_esri_variant_special_cases = [
    ['DATUM["D_OSGB_1936"','DATUM["OSGB_1936"']
];

module.exports.canonical = {
    spherical_mercator: {
        proj4: '+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over',
        srid: 3857, // or 900913
        auth: 'EPSG',
        pretty_wkt: merc_pretty_wkt,
        esri: false,
        name: 'Google Maps Global Mercator', // so many variants here
        valid: true,
        is_geographic: false
    },
    wgs84: {
        proj4: '+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs',
        srid: 4326,
        auth: 'EPSG',
        pretty_wkt: lon_lat_pretty_wkt,
        esri: false,
        name: 'WGS 84',
        valid: true,
        is_geographic: true
    }
};
